client_hooks.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2018 Twitch Interactive, Inc.  All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may not
+// use this file except in compliance with the License. A copy of the License is
+// located at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// or in the "license" file accompanying this file. This file is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+package twirp
+
+import (
+	"context"
+	"net/http"
+)
+
+// ClientOption is a functional option for extending a Twirp client.
+type ClientOption func(*ClientOptions)
+
+// ClientOptions encapsulate the configurable parameters on a Twirp client.
+type ClientOptions struct {
+	Hooks *ClientHooks
```

Comment:
I'm assuming the ServerHooks are implemented in the same way, but you could instead:

- Make this a private `hooks *ClientHooks`
- Make `WithClientHooks` implement the logic in `ChainClientHooks` by automatically chaining if `hooks != nil`
- Have `func (h *ClientHooks) Hooks() *ClientHooks` that the generated code calls

This would simplify the API a bit

Reply:
Ahhhhh, this is a good idea. I'll give this a shot. For the third bullet point, did you mean: `func (o *ClientOptions) Hooks() *ClientHooks`?

Reply:
Addressed.